### PR TITLE
Fix clippy warnings and errors

### PR DESCRIPTION
**Pull request by AngleNet**
Fix warnings and errors while clippy

### CODE REVIEW

**Comment by AngleNet** [comment]
HI, all. Is contributions welcomed for this repo ? I'd like to help ! I fixed some errors.

**Comment by BusyJay** [comment]
> HI, all. Is contributions welcomed for this repo ?

Of course yes! Thanks for your contribution! Can you also add the clippy lint to CI check?

**Comment by AngleNet** [comment]
Add clippy lint in GitHub workflows ?

**Comment by BusyJay** [comment]
Yes.

**Comment by AngleNet** [comment]
Already added the lint. By the way, Why you start to write this key-value store from scratch ? Why no rocksdb ? Could you clear this for me ? Thank you !!!

**Comment by BusyJay** [comment]
Rocksdb is too complicated and general. This project is about to act quickly and bring aggressive optimizations for TiKV only. Writing it in Rust can bring memory safety and be easier integrated into TiKV. This is an experimental engine for TiKV, we will still keep using rocksdb.

**Comment by AngleNet** [comment]
Got it! I looked through the code base. Is there any draft docs over the design for now ? I see that there are only some development plans to implement the core features of a classical key-value store.  Or it's planed to implement the main features for a key-value store firstly and introduce optimizations later ?

**Review by skyzh — APPROVED** [review]
Thanks for your contribution!

File: src/table/builder.rs [hunk 1]
```diff
@@ -242,8 +240,8 @@ mod tests {
 
         let idx = table.inner.read_table_index().unwrap();
 
-        for i in 0..idx.offsets.len() {
-            assert_eq!(block_first_keys[i], idx.offsets[i].key);
+        for (i, k) in block_first_keys.iter().enumerate().take(idx.offsets.len()) {
```
Comment (BusyJay) [inline src/table/builder.rs:243]: What if `block_first_keys.len() < idx.offsets.len()`?

File: src/table/iterator.rs [hunk 2]
```diff
@@ -321,10 +316,7 @@ impl<T: AsRef<TableInner>> Iterator<T> {
         let idx = util::search(self.table.as_ref().offsets_length(), |idx| {
             use std::cmp::Ordering::*;
             let block_offset = self.table.as_ref().offsets(idx).unwrap();
-            match COMPARATOR.compare_key(&block_offset.key, &key) {
-                Greater => true,
-                _ => false,
-            }
+            !matches!(COMPARATOR.compare_key(&block_offset.key, &key), Greater)
```
Comment (BusyJay) [inline src/table/iterator.rs:319]: Redundant `!`.

**Comment by skyzh** [comment]
> Got it! I looked through the code base. Is there any draft docs over the design for now ? I see that there are only some development plans to implement the core features of a classical key-value store. Or it's planed to implement the main features for a key-value store firstly and introduce optimizations later ?

We have updated our roadmap in README.md. Thanks for your interest.

File: src/table/iterator.rs (hunk 2 above)
Comment (AngleNet) [reply src/table/iterator.rs:319]: Oh. Yes

File: src/table/builder.rs (hunk 1 above)
Comment (AngleNet) [reply src/table/builder.rs:243]: Will fix it!

Comment (AngleNet) [reply src/table/builder.rs:243]: @BusyJay  I think it should always be equal. If we didn't get that right. Then the test should fail. Is it correct ?

Comment (skyzh) [reply src/table/builder.rs:243]: I have added real unit tests last week. You may have to rebase your work on latest master so as to see if this change really works.

Comment (skyzh) [reply src/table/builder.rs:243]: I prefer preserving how original code works, by iterating only on offsets.

Comment (AngleNet) [reply src/table/builder.rs:243]: However, the clippy complain about the original code. I'm checking it right now.

Comment (BusyJay) [reply src/table/builder.rs:243]: > I think it should always be equal. If we didn't get that right. Then the test should fail. Is it correct ?

Yes. However after changing to iterator, even they are not equal, the test will not fail. You can either add an explicit assert on length or just dismiss the lint.

Comment (AngleNet) [reply src/table/builder.rs:243]: I prefer to add the assert of length.

Comment (BusyJay) [reply src/table/builder.rs:243]: If they have the same length, then `take(n)` is unnecessary, right? Perhaps you can write `block_first_keys..iter().zip(&idx.offsets)`.

File: src/table/tests.rs [hunk 3]
```diff
@@ -84,7 +84,7 @@ fn test_table_iterator() {
 
 #[test]
 fn test_seek_to_first() {
-    for n in vec![99, 100, 101, 199, 200, 250, 9999, 10000] {
+    for &n in &[99, 100, 101, 199, 200, 250, 9999, 10000] {
```
Comment (BusyJay) [inline src/table/tests.rs:87]: In TiKV, we usually prefer `*` instead of `&`.

File: src/entry.rs [hunk 4]
```diff
@@ -9,6 +10,55 @@ pub struct Entry {
     pub meta: u8,
 }
 
+fn varint_len(l: usize) -> usize {
```
Comment (BusyJay) [inline src/entry.rs:13]: These seem to be deleted.

File: benches/bench_table.rs [hunk 5]
```diff
@@ -28,8 +28,8 @@ fn bench_table_builder(c: &mut Criterion) {
 
         b.iter(|| {
             let mut builder = TableBuilder::new(opt.clone());
-            for j in 0..KEY_COUNT {
-                builder.add(&key_list[j], vs.clone(), 0);
+            for k in key_list.iter().take(KEY_COUNT) {
```
Comment (BusyJay) [inline benches/bench_table.rs:31]: How about `&key_list[..KEY_COUNT]`?

**Comment by skyzh** [comment]
Thanks for your work! Would you please rebase or merge latest master onto this branch? (Or you may open a new PR for clippy). It seems that the git diff contains content from previous commits, and it's hard to review what has really changed.

**Comment by AngleNet** [comment]
> Thanks for your work! Would you please rebase or merge latest master onto this branch? (Or you may open a new PR for clippy). It seems that the git diff contains content from previous commits, and it's hard to review what has really changed.

Of course! I would open another PR for this.

**Pull request closed without merging**